src/Npgsql/TopologyAwareDataSource.cs

```diff
@@ -0,0 +1,289 @@
+                var pref = Convert.ToInt32(v[1]);
+                if (pref == 1)
+                {
+                    HashSet<CloudPlacement>? primary;
+                    if (!allowedPlacements.TryGetValue(PRIMARY_PLACEMENTS, out primary))
+                    {
+                        primary = new HashSet<CloudPlacement>();
+                        allowedPlacements[PRIMARY_PLACEMENTS] = primary;
+                    }
+                    PopulatePlacementSet(v[0], primary);
+                }
+                else if (pref > 1 && pref <= MAX_PREFERENCE_VALUE)
+                {
+                    HashSet<CloudPlacement>? fallbackPlacements;
+                    if (!allowedPlacements.TryGetValue(pref, out fallbackPlacements))
+                    {
+                        fallbackPlacements = new HashSet<CloudPlacement>();
+                        allowedPlacements[pref] = fallbackPlacements;
+                    }
+                    PopulatePlacementSet(v[0], fallbackPlacements);
+                }
+                else
+                {
+                    throw new InvalidExpressionException("Invalid value part for property" + settings.TopologyKeys + ":" + value);
+                }
+            }
+        }
+    }
+
+    void PopulatePlacementSet(string placements, HashSet<CloudPlacement>? allowedPlacements)
+    {
+        var pStrings = placements.Split(',');
+        foreach (var pl in pStrings)
+        {
+            var placementParts = pl.Split('.');
+            if (placementParts.Length != 3 || placementParts[0].Equals("*") || placementParts[1].Equals("*"))
+            {
+                throw new InvalidExpressionException("Malformed " + settings.TopologyKeys + " property value:" + pl);
+            }
+
+            CloudPlacement cp = new CloudPlacement(placementParts[0], placementParts[1], placementParts[2]);
+
+            allowedPlacements?.Add(cp);
+
+        }
+    }
+    
+
+    /// <summary>
+    /// Create a new pool
+    /// </summary>
+    internal new void CreatePool(NpgsqlConnection conn)
+    {
+        _hosts = GetCurrentServers(conn);
+        foreach(var host in _hosts)
+        {
+            var poolSettings = settings.Clone();
+            poolSettings.Host = host.ToString();
+
+            _pools.Add(settings.Pooling
+                ? new PoolingDataSource(poolSettings, dataSourceConfig)
+                : new UnpooledDataSource(poolSettings, dataSourceConfig));
+        }
+    }
+
+    new List<string> GetCurrentServers(NpgsqlConnection conn)
+    {
+        NpgsqlCommand QUERY_SERVER = new NpgsqlCommand("Select * from yb_servers()",conn);
+        NpgsqlDataReader reader = QUERY_SERVER.ExecuteReader();
+        List<string> currentPrivateIps = new List<string>();
+        var hostConnectedTo = conn.Host;
+        
+        Debug.Assert(hostConnectedTo != null, nameof(hostConnectedTo) + " != null");
+        while (reader.Read())
+        {
+            var host = reader.GetString(0);
```

Comment:
Is it possible to read fields by their names instead of index? If yb_servers() is changed in future (as it did in 2.15/2.16 from earlier versions), then this would not break.